contracts/tf20/src/contract.rs

```diff
@@ -0,0 +1,541 @@
+    deps.api.addr_validate(&owner)?;
+    deps.api.addr_validate(&recipient)?;
+
+    _transfer(deps, env, owner, recipient, amount)
+}
+
+pub fn force_burn(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    owner: String,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender)?;
+
+    deps.api.addr_validate(&owner)?;
+    _burn(deps, env, owner, amount)
+}
+
+pub fn force_send(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    owner: String,
+    contract: String,
+    amount: Uint128,
+    msg: Binary,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender)?;
+
+    deps.api.addr_validate(&owner)?;
+    deps.api.addr_validate(&contract)?;
+    _send(deps, env, owner, contract, amount, msg)
+}
+
+pub fn update_contract_admin(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    new_admin: String,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender.clone())?;
+    let old_admin = info.sender.into_string();
+
+    let admin = match new_admin.is_empty() {
+        true => None,
+        false => {
+            let addr = deps.api.addr_validate(&new_admin)?;
+            Some(addr)
+        }
+    };
+
+    let mut token_info = TOKEN_INFO.load(deps.storage)?;
+    token_info.admin = admin;
+    TOKEN_INFO.save(deps.storage, &token_info)?;
+    Ok(Response::new()
+        .add_attribute("action", "update_contract_admin")
+        .add_attribute("old_admin", old_admin)
+        .add_attribute("new_admin", new_admin))
+}
+
+pub fn update_tokenfactory_admin(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    new_admin: String,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender.clone())?;
+    let old_admin = info.sender.clone().into_string();
+
+    let denom = denom(deps.as_ref())?;
+
+    let change_admin_msg = MsgChangeAdmin {
+        sender: info.sender.into_string(),
+        denom,
+        new_admin: new_admin.clone(),
+    };
+
+    let any_msg = AnyMsg {
+        type_url: String::from("/osmosis.tokenfactory.v1beta1.Msg/ChangeAdmin"),
+        value: to_json_binary(&change_admin_msg)?,
+    };
+
+    Ok(Response::new()
+        .add_attribute("action", "update_tokenfactory_admin")
+        .add_attribute("old_admin", old_admin)
+        .add_attribute("new_admin", new_admin)
+        .add_message(CosmosMsg::Any(any_msg)))
+}
+
+pub fn modify_metadata(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    metadata: Binary,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender.clone())?;
+
+    let deserialized_metadata = from_json(metadata)?;
+
+    let change_admin_msg = MsgSetDenomMetadata {
+        sender: info.sender.into_string(),
+        metadata: Some(deserialized_metadata),
+    };
+
+    let any_msg = AnyMsg {
+        type_url: String::from("/osmosis.tokenfactory.v1beta1.Msg/ChangeAdmin"),
```

Comment:
update: SetDenomMetadata, not change admin for meta data modification

Reply:
thanks for the catch